versiondb/client/verify.go

```diff
@@ -45,87 +61,218 @@ func VerifyChangeSetCmd() *cobra.Command {
+	entries, err := os.ReadDir(storeDir)
+	if err != nil {
+		return nil, err
+	}
+	fileNames := make([]string, len(entries))
+	for i, entry := range entries {
+		fileNames[i] = filepath.Join(storeDir, entry.Name())
+	}
+
+	filesWithVersion, err := SortFilesByFirstVerson(fileNames)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(filesWithVersion) == 0 {
+		return nil, fmt.Errorf("change set directory is empty")
+	}
+	// the initial version for the store
+	initialVersion := filesWithVersion[0].Version
+
+	var (
+		tree *memiavl.Tree
+	)
+	if len(loadSnapshot) > 0 {
+		path := filepath.Join(loadSnapshot, store)
+		snapshot, err := memiavl.OpenSnapshot(path)
+		if err != nil {
+			return nil, errors.Wrapf(err, "fail to load snapshot: %s", path)
+		}
+		defer snapshot.Close()
+
+		tree = memiavl.NewFromSnapshot(snapshot)
+		fmt.Printf("snapshot loaded: %d %X\n", tree.Version(), tree.RootHash())
+	} else {
+		tree = memiavl.NewWithInitialVersion(int64(initialVersion))
+	}
+
+	for _, file := range filesWithVersion {
+		if targetVersion > 0 && file.Version > uint64(targetVersion) {
+			break
+		}
+
+		err = withChangeSetFile(file.FileName, func(reader Reader) error {
+			_, err := IterateChangeSets(reader, func(version int64, changeSet *iavl.ChangeSet) (bool, error) {
+				if version <= tree.Version() {
+					// skip old change sets
+					return true, nil
+				}
+
+				for _, pair := range changeSet.Pairs {
+					if pair.Delete {
+						tree.Remove(pair.Key)
+					} else {
+						tree.Set(pair.Key, pair.Value)
+					}
+				}
+
+				// no need to update hashes for intermediate versions.
+				_, v, err := tree.SaveVersion(false)
+				if err != nil {
+					return false, err
+				}
+				if v != version {
+					return false, fmt.Errorf("version don't match: %d != %d", v, version)
+				}
+				return targetVersion == 0 || v < targetVersion, nil
+			})
+
+			return err
+		})
+
+		if err != nil {
+			break
+		}
+
+		if targetVersion > 0 && tree.Version() >= targetVersion {
+			break
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(saveSnapshot) > 0 {
+		snapshotDir := filepath.Join(saveSnapshot, store)
+		if err := os.MkdirAll(snapshotDir, os.ModePerm); err != nil {
+			return nil, err
+		}
+		if err := tree.WriteSnapshot(snapshotDir); err != nil {
+			return nil, err
+		}
+	}
+
+	return &storetypes.StoreInfo{
+		Name:     store,
+		CommitId: lastCommitID(tree),
+	}, nil
+}
+
+// lastCommitID build `CommitID` from a memiavl tree.
+func lastCommitID(tree *memiavl.Tree) storetypes.CommitID {
+	// copy out the hash in case it's relied on mmap-ed file.
+	var hash [memiavl.SizeHash]byte
+	copy(hash[:], tree.RootHash())
+	return storetypes.CommitID{
+		Version: tree.Version(),
+		Hash:    hash[:],
+	}
+}
+
+// buildCommitInfo sort the storeInfos by store name, and built `CommitInfo`.
+func buildCommitInfo(storeInfos []storetypes.StoreInfo, version int64) storetypes.CommitInfo {
+	sort.SliceStable(storeInfos, func(i, j int) bool {
```

Comment:
not sure if just align like sdk's [sort.Slice](https://github.com/cosmos/cosmos-sdk/blob/main/store/rootmulti/store.go#L47), which is also faster

Reply:
https://github.com/cosmos/cosmos-sdk/blob/main/store/rootmulti/store.go#L1098
i actually copied from sdk